### PR TITLE
tdigest sketch support for aggregation and post aggregation

### PR DESCRIPTION
**Pull request by tannaga**
PR for #25

### CODE REVIEW

**Comment by tannaga** [comment]
@vigith @jbguerraz Please review the PR. Also I would like to get your thoughts on the below item

Druid returns merged tdigest data in a base64 encoded format. To convert it to actual TDigest, we need to write a custom helper method which should more or less mimics the [logic in Java used in Druid](https://github.com/tdunning/t-digest/blob/main/core/src/main/java/com/tdunning/math/stats/MergingDigest.java#L899).

Decoding the data is helpful to plot histograms/heatmap

Question: 
1. Should this be part of this repo in a helper package or similar?
2. I am converting it to a [caio/tdigest](https://github.com/caio/go-tdigest) instance in this code snippet, but i could as well return a custom struct like 
      ```go
            type TDigestData {
                 mean []float64
                 weight []uint64
            }
      ```

Here is the code snippet
```go
func FromBytes(buf *bytes.Reader) (*tdigest.TDigest, error) {
	var encoding int32
	var endianess = binary.BigEndian
	log.Printf("buffer length: %d", buf.Len())
	err := binary.Read(buf, endianess, &encoding)
	if err != nil {
		return nil, err
	}
	log.Printf("buffer length: %d", buf.Len())
	if encoding != 1 {
		return nil, fmt.Errorf("Unsupported encoding version: %d", encoding)
	}
	var min float64
	err = binary.Read(buf, endianess, &min)
	if err != nil {
		return nil, err
	}
	var max float64
	err = binary.Read(buf, endianess, &max)
	if err != nil {
		return nil, err
	}
	var compression float64
	err = binary.Read(buf, endianess, &compression)
	if err != nil {
		return nil, err
	}
	td, err := tdigest.New(tdigest.Compression(compression))
	var index uint32
	err = binary.Read(buf, endianess, &index)
	if err != nil {
		return nil, err
	}
	for i := 0; i < int(index); i++ {
		var weight float64
		err = binary.Read(buf, endianess, &weight)
		if err != nil {
			return nil, err
		}

		var mean float64
		err = binary.Read(buf, endianess, &mean)
		if err != nil {
			return nil, err
		}
		td.AddWeighted(mean, uint64(weight))
	}
	log.Printf("compression: %f", compression)
	return td, nil
}
```

File: examples/tdigestExample.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+		log.Fatal(err)
+	}
+	fmt.Println("{\"version\": \"" + status.Version + "\"}")
+	return d
+}
+
+// tdigestSketchUsingBuilder example using Builder Pattern
+func tdigestSketchUsingBuilder() {
+	d := getConnection()
+	table := datasource.NewTable().SetName("rollup-data")
```
Comment (vigith) [inline examples/tdigestExample.go:34]: could you please add documentation on how to create this data-source?

**Comment by vigith** [comment]
I am not sure what @jbguerraz has in mind for the roadmap of go-druid. If we are to keep go-druid as a builder pattern for druid native querying (runes), I believe we should keep parsing the tdigest outside of repo.

Today the output of each query type is returned as an interface and it is left to the client to unpack.

File: examples/tdigestExample.go (hunk 1 above)
Comment (tannaga) [reply examples/tdigestExample.go:40]: done

**Review by jbguerraz — CHANGES_REQUESTED** [review]

**Comment by jbguerraz** [comment]
Thank you @tannaga ! This looks great :+1: 

> I am not sure what @jbguerraz has in mind for the roadmap of go-druid. If we are to keep go-druid as a builder pattern for druid native querying (runes), I believe we should keep parsing the tdigest outside of repo.
> 
> Today the output of each query type is returned as an interface and it is left to the client to unpack.

This is a very good point. I'd say `go-druid` is a wrapper for Druid APIs (there is already common APIs covered, I'd be happy if end up covering all of them: https://druid.apache.org/docs/latest/operations/api-reference.html ).

About tdigest, I'd be happy if we could use it to render histograms/heatmaps on Grafana, like https://grafana.com/docs/grafana/latest/getting-started/intro-histograms/ and for doing so the backend side (golang) of the plugin  (https://github.com/grafadruid/druid-grafana/blob/master/pkg/druid.go) would need to decode the tdigest.

Not sure neither where this should be done, it's actually pretty specific and I dunno what other specific things could bring other extensions in future.

We could create a new repo `go-druid-tdigestsketch` in this `grafadruid` organization to do just this decoding or go with a specific decoder stored in some `/decoders` directory at the root of `go-druid` repository (`/decoders/tdigestsketch.go` or `/decoders/tdigestsketch/tdigestsketch.go`) ? WDYT @tannaga  @vigith (or anyone reading this ;)) ?

**Comment by vigith** [comment]
IMHO, let's create, say `go-druid-extensions` and add all these use cases there (eg. pkg called `tdigest` in `go-druid-extensions`). Adding it in `go-druid` will makes it very difficult to call `go-druid` feature complete since there are a lot of extensions out there. WDYT?

**Comment by jbguerraz** [comment]
I like the proposal of keeping extensions separated @vigith :+1: 
Eager to get your POV @tannaga !

**Comment by tannaga** [comment]
+1  on Vigith's proposal. 
Would it make sense to move the aggregation and post aggregation code as well to that repo as shown below? . In that way it aligns with the way druid manages extensions and also consumers can pull in specific extensions on need basis

```
go-druid-extensions
└── tdigest
    ├── aggregation
    ├── decoder
    ├── go.mod
    ├── go.sum
    └── postaggregation

4 directories, 2 files
```
If we go down this path then how do we handle this part of the code in aggregation.go?
```
	case "tDigestSketch":
		a = NewTDigestSketch()		
	}
```

**Comment by jbguerraz** [comment]
@tannaga @vigith maybe we could profit a component registry here? something like `aggregation.registerAggregator(type, factory)` (e.g: `aggregation.registerAggregator('tDigestSketch', NewTDigestSketch)`) maybe called from `init` method ?

**Comment by vigith** [comment]
I think we can put `aggregation` and `postaggregation` into `go-druid`. The decoding (encoding, etc) can stay in a different module which is used only by the client code to decode once data is retrieved from druid. 
```
go-druid-extensions/
└── tdigest
    ├── decoder
    │   └── files.go
    ├── encoder
    │   └── files.go
    ├── go.mod
    └── gosum
```

I do not think we should include these extensions into `go-druid`'s `go.mod` as it might bloat up the dependencies. If we use component registry then there will be a dependency as `init` will have to reference the module.

WDYT?

**Comment by tannaga** [comment]
@vigith I was initially thinking on those lines. So can we say `go-druid` as feature complete even though we have not covered all the extensions?

Since aggregation/post aggregation does not add any new third party library for some of the extensions I have seen, we could probably leave them in the `go-druid` repo

**Comment by vigith** [comment]
> @vigith I was initially thinking on those lines. So can we say `go-druid` as feature complete even though we have not covered all the extensions?
> 
 Yes, it is easy to add those to `go-druid` compared to decoders and encoders.

**Comment by tannaga** [comment]
I am onboard with the approach!
If all agree, then this PR can stay as is and when the new extension repo is created, i can add the encode/decode functionality.

**Review by jbguerraz — APPROVED** [review]
I'm still not yet fully confident about how to manage extensions but at this point I believe it worth it to move ahead.

**Comment by jbguerraz** [comment]
LGTM :) Thank you @tannaga !!

**Pull request merged**